Add .expand() method to distribution classes

neerajprad opened the pull request:
This adds a `.expand` method for distributions that is akin to the `torch.Tensor.expand` method for tensors. It returns a new distribution instance with batch dimensions expanded to the desired `batch_shape`. Since this calls `torch.Tensor.expand` on the distribution's parameters, it does not allocate new memory for the expanded distribution instance's parameters. 

e.g.
```python
>>> d = dist.Normal(torch.zeros(100, 1), torch.ones(100, 1))
>>> d.sample().shape
  torch.Size([100, 1])
>>> d.expand([100, 10]).sample().shape
  torch.Size([100, 10])
```

### Motivation

We have already been using the `.expand` method in Pyro in our [patch](https://github.com/uber/pyro/blob/dev/pyro/distributions/torch.py#L10) of `torch.distributions`. We use this in our models to enable dynamic broadcasting. This has also been requested by a few users on the distributions slack, and we believe will be useful to the larger community. 

Note that currently, there is no convenient and efficient way to expand distribution instances:
 - Many distributions use `TransformedDistribution` (or wrap over another distribution instance. e.g. `OneHotCategorical` uses a `Categorical` instance) under the hood, or have lazy parameters. This makes it difficult to collect all the relevant parameters, broadcast them and construct new instances.
 - In the few cases where this is even possible, the resulting implementation would be inefficient since we will go through a lot of broadcasting and args validation logic in `__init__.py` that can be avoided.

The `.expand` method allows for a safe and efficient way to expand distribution instances. Additionally, this bypasses `__init__.py` (using `__new__` and populating relevant attributes) since we do not need to do any broadcasting or args validation (which was already done when the instance was first created). This can result in significant savings as compared to constructing new instances via `__init__` (that said, the `sample` and `log_prob` methods will probably be the rate determining steps in many applications).

e.g.
```python
>>> a = dist.Bernoulli(torch.ones([10000, 1]), validate_args=True)

>>> %timeit a.expand([10000, 100])
15.2 µs ± 224 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

>>> %timeit dist.Bernoulli(torch.ones([10000, 100]), validate_args=True)
11.8 ms ± 153 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

cc. @fritzo, @apaszke, @vishwakftw, @alicanb

Commit:
Add .expand() method to distribution classes

Commit:
remove default arg

Commit:
uncomment run tests

Commit:
fix lint error

Commit:
remove default arg

vishwakftw commented on test/test_distributions.py:
```diff
@@ -791,6 +791,24 @@ def test_has_examples(self):
+    def test_distribution_expand(self):
+        shapes = [torch.Size(), torch.Size((2,)), torch.Size((2, 1))]
+        for Dist, params in EXAMPLES:
+            if Dist.__name__ == "TransformedDistribution":
```
Cannot this be `isinstance(Dist, TransformedDistribution)` ?

fritzo replied:
We subclass `TransformedDistribution` in many cases, and I believe Neeraj is ensuring that those subclasses are tested.

vishwakftw commented on test/test_distributions.py:
```diff
@@ -2175,6 +2193,25 @@ def test_independent_shape(self):
 
+    def test_independent_expand(self):
+        for Dist, params in EXAMPLES:
+            if Dist.__name__ == "TransformedDistribution":
```
Same here.

vishwakftw commented on test/test_distributions.py:
```diff
@@ -2175,6 +2193,25 @@ def test_independent_shape(self):
+        for Dist, params in EXAMPLES:
+            if Dist.__name__ == "TransformedDistribution":
+                continue
+            for i, param in enumerate(params):
```
Why is `enumerate` required?

vishwakftw commented on torch/distributions/bernoulli.py:
```diff
@@ -46,6 +46,19 @@ def __init__(self, probs=None, logits=None, validate_args=None):
+        else:
+            new.logits = self.logits.expand(batch_shape)
+            new._param = new.logits
+        super(Bernoulli, new).__init__(batch_shape, validate_args=False)
```
Why can't `validate_args` be `self.validate_args`? And is `validate_args` actually required? Because, if `validate_args` is true in the parent distribution, then the arguments are validated there, and this is merely an expanded version. Sorry if I missed something.

fritzo commented on torch/distributions/gumbel.py:
```diff
@@ -31,14 +31,20 @@ def __init__(self, loc, scale, validate_args=None):
         if isinstance(loc, Number) and isinstance(scale, Number):
-            batch_shape = torch.Size()
-            base_dist = Uniform(finfo.tiny, 1 - finfo.eps)
+            self._base_dist = Uniform(finfo.tiny, 1 - finfo.eps)
```
Can you explain the pattern in this PR of using new attributes `._base_dist, ._transform` rather than the existing `.base_dist,.transform` attributes? I don't see why the existing attributes can be used instead.

neerajprad replied:
My bad; I missed that these were already available as these are subclassing `TransformedDistribution`. Will remove.

neerajprad commented on torch/distributions/bernoulli.py:
```diff
@@ -46,6 +46,19 @@ def __init__(self, probs=None, logits=None, validate_args=None):
 
+    def expand(self, batch_shape):
+        batch_shape = torch.Size(batch_shape)
+        new = self.__new__(Bernoulli)
```
That is unfortunately true. :( I think we can use `type(self)` to ensure that we are getting the correct instance type, but if we have made any customizations to `__init__`, this will require us to also override `.expand` and plug in those changes. That's not ideal. Another thing we can do is check if `type(self)` doesn't match the class and if so call the `__init__` method for safety.

fritzo replied:
Hmm, I think it's too draconian to `assert type(self) is Bernoulli`, since we should allow subclassing w/o overriding the `__init__` method. I think the safest thing to do would be:
```py
if self.__init__.im_func is not Bernoulli.__init__.im_func:
    raise NotImplementedError("Subclasses that define a custom __init__ method "
                              "must also define a custom .expand() method")
```

neerajprad replied:
Yes, that would be better. Rather than trying to do something by default, it's better to throw this error since its not currently possible to `.expand` without making any assumptions about `.__init__` (e.g. we may introduce some new parameters that would additionally require broadcasting).

fritzo replied:
The `type(self).__init__ is not Bernoulli.__init__` version seems palatable.

neerajprad replied:
Hmm..I think this is getting a bit complicated. Consider a distribution `X` that subclasses `Y`. We may call `super(X, self).__init__(...)` to set attributes defined in the parent class. If we want to use `super(X, self).expand(...)` to expand those attributes, this isn't currently possible (apart from this additional check) since `.expand` not just sets the attributes (like `__init__`) but also creates a new instance via `__new__`.  One way (there may be a better one) might be to take in an optional `instance` arg to enable subclasses to use the `expand` method :

```
    def expand(self, batch_shape, instance=None):
        if not instance and type(self).__init__ is not Bernoulli.__init__:
            raise NotImplementedError("Subclasses that define a custom __init__ method "
                                      "must also define a custom .expand() method")
        new = self.__new__(Bernoulli) if not instance else instance
        ...
```

fritzo replied:
I don't think we need to support generic `.expand()` via `super(X, self).expand()`. In cases where a user wants to subclass with an `__init__` and use `.expand()`, I guess we could require the user to hand-code an `.expand()` method? That's what we're doing with `TransformedDistribution`, right?

neerajprad replied:
We are still able to call `base_dist.expand()` in `TransformedDistribution`, so it isn't as bad. For subclasses, we will need to duplicate the `.expand` code from the parent class just to set those attributes. e.g. in `Chi2`, we will need to set `concentration` and `rate` in `.expand`, which are not used inside `Chi2`. I'm not sure whether we should provide this extra `instance` kwarg for convenience, or leave it to the users to hand-code it.

Commit:
address review comments

Commit:
add an instance attribute for subclasses

Commit:
rearrange

Commit:
fix error on python 2.7

Commit:
add subclass expand test

Commit:
make method compatible across py 2.7/3

neerajprad commented:
I think this addresses all comments, and is ready for another review. 

One thing that is worth doing as suggested by @fritzo (either here or in a separate PR) is to call `.contiguous()` inside `.expand()` which will result in faster `.sample()` and `.log_prob()`. I am debating whether we should do this by default, or provide another kwarg `contiguous=False`.

vishwakftw commented:
@neerajprad Calling `.contiguous()` will probably not satisfy the objective of this PR, because of the fact that `.contiguous()` creates a contiguous copy of the tensor, meaning more memory allocation, which is not intended if I understand correctly.

fritzo commented on torch/distributions/distribution.py:
```diff
@@ -220,6 +241,19 @@ def _validate_sample(self, value):
+            """
+            return getattr(cls_.__init__, "__func__", cls_.__init__)
+
+        if instance is None and cons(type(self)) is not cons(cls):
```
You can simplify to `type(self).__init__ == cls.__init__`, which appears to work in both Python 2.7 and 3.

neerajprad replied:
You are right! Equality works fine actually (they have the same id), but `is` fails.

neerajprad commented:
>  Calling .contiguous() will probably not satisfy the objective of this PR, because of the fact that .contiguous() creates a contiguous copy of the tensor, meaning more memory allocation, which is not intended if I understand correctly.

@vishwakftw - We can support `Distribution.expand` by either reusing existing params or creating new ones altogether (even this should be faster than going through __init__), though it is preferable to not allocate additional memory whenever possible. `.contiguous()` will do exactly that - it will not create a new copy if the existing one is already contiguous. In many applications, memory may not be an issue, but the higher time taken to draw or score samples from expanded instances might be (specially as it completely avoidable). e.g. if we expand an instance and are calling `.sample` multiple times. In such cases, the overhead of using non contiguous parameters is high enough that it is better to pay a one time cost of calling `.contiguous`. I am happy to defer this to a separate PR / discuss in another issue though.

fritzo commented on torch/distributions/distribution.py:
```diff
@@ -35,6 +35,27 @@ def __init__(self, batch_shape=torch.Size(), event_shape=torch.Size(), validate_
                 if not constraint.check(getattr(self, param)).all():
                     raise ValueError("The parameter {} has invalid values".format(param))
 
+    def expand(self, batch_shape, instance=None):
```
Consider renaming `instance` to `_instance` and document generally only used by derived classes that need to override the `.expand()` method. It's kind of gross, but I don't see a simpler solution.

fritzo commented on test/test_distributions.py:
```diff
@@ -791,6 +791,47 @@ def test_has_examples(self):
+                    self.assertEqual(expanded.log_prob(sample), d.log_prob(sample))
+                    self.assertEqual(actual_shape, expected_shape)
+
+    def test_distribution_subclass_expand(self):
```
:+1:

fritzo commented on torch/distributions/distribution.py:
```diff
@@ -220,6 +241,19 @@ def _validate_sample(self, value):
+            raise NotImplementedError("Subclass {} of {} that defines a custom __init__ method "
+                                      "must also define a custom .expand() method.".
+                                      format(self.__class__.__name__, cls.__name__))
+        return self.__new__(type(self)) if not instance else instance
```
It's safer to `... if instance is None else instance`, in case someone defines a `__bool__` method.

fritzo commented:
> the higher time taken to draw or score samples from expanded instances

Can you either create an issue describing this performance bug in `torch.bernoulli(x.expand(...))`, or point to an existing issue to provide context? I think @vishwakftw and I are is missing context of timing experiments you've performed regarding `.contiguous()`.

neerajprad commented:
> Can you either create an issue describing this performance bug in torch.bernoulli(x.expand(...)), or point to an existing issue to provide context? I think @vishwakftw and I are is missing context of timing experiments you've performed regarding .contiguous().

Will create an issue, and point to it. It is probably not a bug though, and may just be due to the overhead of having to deal with non contiguous parameters.

Commit:
address review comments

neerajprad commented:
@vishwakftw, @fritzo : I opened a separate issue for discussion here - https://github.com/pytorch/pytorch/issues/11389.

fritzo reviewed (approved):
LGTM, looks great!

I'm looking forward to cheaper `.expand()` methods in Pyro (and to the deletion of `ReshapedDistribution`).

neerajprad commented:
Thanks for reviewing, @fritzo, @vishwakftw!

neerajprad commented:
@soumith - Can we get this merged into master? I would like to test the integration on the Pyro side once it is landed.

soumith commented:
sure, i'm merging it now.
